zerver/actions/message_delete.py

```diff
@@ -153,3 +160,97 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
+    This prevents issues where events are sent to wrong users when processing
+    messages from different directions of the same conversation separately.
+
+    Returns:
+        - Dictionary mapping other participant ID to their 1:1 messages with sender
+        - List of group DM messages (which can be processed normally)
+    """
+    personal_messages_by_other_user: dict[int, list[Message]] = {}
+    direct_message_group_messages: list[Message] = []
+
+    for message in user_direct_messages:
+        if message.recipient.type == Recipient.PERSONAL:
+            # For 1:1 DMs, group by the other participant
+            # message.recipient.type_id is the recipient user ID
+            other_user_id = message.recipient.type_id
+            if other_user_id not in personal_messages_by_other_user:
+                personal_messages_by_other_user[other_user_id] = []
+            personal_messages_by_other_user[other_user_id].append(message)
+        else:
+            # For group DMs, we can process them normally since they share the same recipient
+            direct_message_group_messages.append(message)
+
+    return personal_messages_by_other_user, direct_message_group_messages
+
+
+def delete_deactivated_user_messages(
+    realm: Realm, user_profile: UserProfile, message_delete_action: MessageDeleteAction
+) -> None:
+    if not (
+        message_delete_action.get("delete_public_stream_messages", False)
+        or message_delete_action.get("delete_private_stream_messages", False)
+        or message_delete_action.get("delete_direct_messages", False)
+    ):
+        return
+
+    if message_delete_action.get(
+        "delete_public_stream_messages", False
+    ) or message_delete_action.get("delete_private_stream_messages", False):
+        user_stream_messages = list(
+            Message.objects.filter(
+                realm_id=user_profile.realm_id,
+                sender=user_profile,
+                recipient__type=Recipient.STREAM,
+            ).select_related("recipient")
+        )
+
+        messages_to_delete = []
+        if message_delete_action.get(
+            "delete_public_stream_messages", False
+        ) and message_delete_action.get("delete_private_stream_messages", False):
+            # Delete all stream messages
+            messages_to_delete = user_stream_messages
+        else:
+            # Only delete specific types
+            private_stream_messages, public_stream_messages = classify_stream_messages(
+                realm, user_stream_messages
+            )
+            if message_delete_action.get("delete_public_stream_messages", False):
+                messages_to_delete.extend(public_stream_messages)
+            if message_delete_action.get("delete_private_stream_messages", False):
+                messages_to_delete.extend(private_stream_messages)
+
+        if messages_to_delete:
+            do_delete_messages(user_profile.realm, messages_to_delete, acting_user=user_profile)
```

Comment:
I think `acting_user` here should be set as user deactivating the user not the user being deactivated.